internal/util/idlocker.go

```diff
@@ -59,3 +61,224 @@ func (vl *VolumeLocks) Release(volumeID string) {
+	return ol.tryAcquire(createOp, volumeID)
+}
+
+// GetCloneLock gets the clone lock on given volumeID.
+func (ol *OperationLock) GetCloneLock(volumeID string) error {
+	return ol.tryAcquire(cloneOpt, volumeID)
+}
+
+// GetDeleteLock gets the delete lock on given volumeID,ensures that there is
+// no clone,restore and expand operation on given volumeID.
+func (ol *OperationLock) GetDeleteLock(volumeID string) error {
+	return ol.tryAcquire(deleteOp, volumeID)
+}
+
+// GetRestoreLock gets the restore lock on given volumeID,ensures that there is
+// no delete operation on given volumeID.
+func (ol *OperationLock) GetRestoreLock(volumeID string) error {
+	return ol.tryAcquire(restoreOp, volumeID)
+}
+
+// GetExpandLock gets the expand lock on given volumeID,ensures that there is
+// no delete and clone operation on given volumeID.
+func (ol *OperationLock) GetExpandLock(volumeID string) error {
+	return ol.tryAcquire(expandOp, volumeID)
+}
+
+// ReleaseSnapshotCreateLock releases the create lock on given volumeID.
+func (ol *OperationLock) ReleaseSnapshotCreateLock(volumeID string) {
+	ol.release(createOp, volumeID)
+}
+
+// ReleaseCloneLock releases the clone lock on given volumeID.
+func (ol *OperationLock) ReleaseCloneLock(volumeID string) {
+	ol.release(cloneOpt, volumeID)
+}
+
+// ReleaseDeleteLock releases the delete lock on given volumeID.
+func (ol *OperationLock) ReleaseDeleteLock(volumeID string) {
+	ol.release(deleteOp, volumeID)
+}
+
+// ReleaseRestoreLock releases the restore lock on given volumeID.
+func (ol *OperationLock) ReleaseRestoreLock(volumeID string) {
+	ol.release(restoreOp, volumeID)
+}
+
+// ReleaseExpandLock releases the expand lock on given volumeID.
+func (ol *OperationLock) ReleaseExpandLock(volumeID string) {
+	ol.release(expandOp, volumeID)
+}
+
+// release deletes the lock on volumeID.
+func (ol *OperationLock) release(op operation, volumeID string) {
+	ol.mux.Lock()
+	defer ol.mux.Unlock()
+	switch op {
+	case cloneOpt:
+		if val, ok := ol.locks[cloneOpt][volumeID]; ok {
+			// decrement the counter for clone operation
+			if val > 1 {
+				ol.locks[cloneOpt][volumeID] = val - 1
+			} else {
+				delete(ol.locks[cloneOpt], volumeID)
+			}
+		}
+	case createOp:
+		if val, ok := ol.locks[createOp][volumeID]; ok {
+			// decrement the counter for create operation
+			if val > 1 {
+				ol.locks[createOp][volumeID] = val - 1
+			} else {
+				delete(ol.locks[createOp], volumeID)
+			}
+		}
+	case deleteOp:
+		// remove volumeID from map
+		delete(ol.locks[deleteOp], volumeID)
+	case restoreOp:
+		if val, ok := ol.locks[restoreOp][volumeID]; ok {
+			if val > 1 {
+				// decrement the counter for restore operation
+				ol.locks[restoreOp][volumeID] = val - 1
+			} else {
+				delete(ol.locks[restoreOp], volumeID)
+			}
+		}
+	case expandOp:
+		// remove volumeID from map
+		delete(ol.locks[expandOp], volumeID)
+	default:
+		klog.Errorf("%v operation not supported", op)
+	}
```

Comment:
It would be shorter to just do the `val - 1` for everything, and then a single `if val == 0 { delete(..) }` before returning from this function. No need to have the `val > 1` check for each case.

Reply:
I meant that `delete()` is not needed in every case. Just return in the `default:` 

Remove these from the cases
```
			if ol.locks[cloneOpt][volumeID] == 0 {
				delete(ol.locks[cloneOpt], volumeID)
			}
```
Only have the subtraction there (for each case?).

And before the end of the function you can do
```
	if ol.locks[op][volumeID] == 0 {
		delete(ol.locks[op], volumeID)
	}
```

This reduces a number of if-statements, and no special handling is needed for `expandOp` either. You could even use `fallthrough` and have a single handler for all operations that way.